tools/pybug.py

```diff
@@ -0,0 +1,1544 @@
+    "app_dump": (
+        cmd_app_dump,
+        "",
+        "Show app data for this chip"),
+    "cmd": (
+        cmd_cmd,
+        '<cmd.D> <arg1.X> <arg2.X> <arg3.X>',
+        'User specified command'),
+}
+
+# ------------------------------------------------------------------------------
+
+
+def usage():
+    print("usage: pybug <options> <hostname>", file=sys.stderr)
+    print("  -bmp  <name>[/<slots>]   - set BMP target", file=sys.stderr)
+    print("  -version                 - print version number", file=sys.stderr)
+    print("  -norl                    - don't use 'ReadLine'", file=sys.stderr)
+    print("  -expert                  - set 'expert' mode", file=sys.stderr)
+    print("  -debug <value>           - set debug variable", file=sys.stderr)
+    sys.exit(1)
+
+
+def process_args():
+    global spinn_target, bmp_target, expert, _readline, debug
+    global bmp_range
+    _range = "0"
+    args = list(sys.argv)
+    while args:
+        arg = args.pop(0)
+        if arg == "-bmp":
+            bmp_target = args.pop(0)
+            m = re.match(r"^(.*)/(\S+)$", bmp_target)
+            if m:
+                bmp_target, _range = m.groups()
+        elif arg == "-version":
+            cmd_version(None)
+            sys.exit()
+        elif arg == "-debug":
+            debug = int(args.pop(0))
+        elif arg == "-norl":
+            _readline = False
+        elif arg == "-expert":
+            expert = True
+        elif not re.match(r"^-", arg):
+            spinn_target = arg
+            break
+        else:
+            usage()
+    if args:
+        usage()
+
+    bmp_range = parse_bits(_range, 0, 23)
+    if spinn_target is None:
+        print("target not specified", file=sys.stderr)
+        sys.exit(1)
+    prompt = spinn_target
+    if not re.match(r"^\d", prompt):
+        prompt = re.sub(r"\..+", "", prompt)
+    prompt += ":0,0,0 > "
+    return prompt
+
+
+def open_targets():
+    global spin, sv, bmp
+    spin = SCAMPCmd(target=spinn_target, port=spin_port, debug=debug)
+    sv = Struct(scp=spin)
+    if bmp_target is not None:
+        bmp = BMPCmd(target=bmp_target, port=bmp_port, debug=debug)
+
+
+class Completer(object):
+    def __init__(self, rl):
+        self._rl = rl
+        self._stored = [None]
+
+    def __call__(self, text, state):
+        if self._rl.get_begidx():
+            # No filename completion; that's complicated!
+            return None
+        if not state:
+            # Add None to the end to mark end of matches
+            # https://eli.thegreenplace.net/2016/basics-of-using-the-readline-library/
+            self._stored = list(_cli.commands_starting_with(text)) + [None]
+        word = self._stored[state]
+        if word and len(self._stored) == 1:
+            word += " "
+        return word
+
+
+def init_readline():
+    if not _readline:
+        return None
+
+    import readline
+    readline.set_completer(Completer(readline))
+    return readline
+
+
+def main():
```

Comment:
Needs minimal documentation on how to run.

This can/or should be limited to how to reach the internal help.

Equivellent of ybug -help

Reply:
Added